### ISSUE TITLE
partition' fails for various files with error: PARTITION: wfGroupNothing

### ISSUE THREAD

**Issue by christetreault** [ucsd-progsys/liquid-fixpoint#88]
For the following liquidhaskell test suite files:

* pos/Holes.hs
* pos/absref-crash0.hs
* pos/alphaconvert-List.hs
* pos/LocalTermExpr.hs
* pos/LocalHole.hs

The call to `partition'` is throwing this exception:

```
**** ERROR *********************************************************************
**** ERROR *********************************************************************
**** ERROR *********************************************************************
PARTITION: wfGroupNothing
**** ERROR *********************************************************************
**** ERROR *********************************************************************
**** ERROR *********************************************************************

```

This behavior can be reproduced using the latest build from the cores branch. These files will return Sat using the singlethreaded solver `liquid [file].hs` but will throw when calling evaluating the list returned by `partition'` when solving with multiple threads `liquid --cores=4 [file].hs`

**Comment by ranjitjhala** [comment]
thanks, I'll take a look at this tomorrow! (Are you blocked on this? If so,
feel free to switch to the book.)

On Tue, Jul 28, 2015 at 3:14 PM, Chris Tetreault <notifications@github.com>
wrote:

> For the following liquidhaskell test suite files:
>
>    - pos/Holes.hs
>    - pos/absref-crash0.hs
>    - pos/LocalTermExpr.hs
>    - pos/LocalHole.hs
>
> The call to partition' is throwing this exception:
>
> **** ERROR *********************************************************************
> **** ERROR *********************************************************************
> **** ERROR *********************************************************************
> PARTITION: wfGroupNothing
> **** ERROR *********************************************************************
> **** ERROR *********************************************************************
> **** ERROR *********************************************************************
>
>
> This behavior can be reproduced using the latest build from the cores
> branch. These files will return Sat using the singlethreaded solver liquid
> [file].hs but will throw when calling evaluating the list returned by
> partition' when solving with multiple threads liquid --cores=4 [file].hs
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquid-fixpoint/issues/88>.
>



-- 
Ranjit.

**Comment by christetreault** [comment]
I have stuff to work on, but that's always been my plan when I reach a roadblock.

BTW, is it normal that a bunch of test are failing, even in the master branch? I'm getting the following failures in the test suite:

```
Tests
  Unit
    pos
      Map2.hs:                    FAIL (2.65s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      alphaconvert-List.hs:       FAIL (1.04s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      zipper.hs:                  FAIL (1.60s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      inline.hs:                  FAIL (0.82s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      MeasureDups.hs:             FAIL (0.83s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      RBTree.hs:                  FAIL (1.32s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      Solver.hs:                  FAIL (1.38s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      take.hs:                    FAIL (0.92s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      AVL.hs:                     FAIL (1.44s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      RBTree-col-height.hs:       FAIL (2.83s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      zipper0.hs:                 FAIL (0.95s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      AVLRJ.hs:                   FAIL (1.86s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      RBTree-height.hs:           FAIL (2.28s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
      ite1.hs:                    FAIL (0.69s)
        Wrong exit code
        expected: ExitSuccess
         but got: ExitFailure 2
    neg
      MeasureDups.hs:             FAIL (0.83s)
        Wrong exit code
        expected: ExitFailure 1
         but got: ExitFailure 2
      Solver.hs:                  FAIL (1.38s)
        Wrong exit code
        expected: ExitFailure 1
         but got: ExitFailure 2
```

**Comment by ranjitjhala** [comment]
No all tests should be passing in master...

**Comment by christetreault** [comment]
I just pulled liquidhaskell and liquid-fixpoint, ensured I had master checked out, and blew away and re-initialized my cabal sandboxes. I'm still getting these same failures.

Perhaps I should open a separate issue for this in liquidhaskell?

**Comment by ranjitjhala** [comment]
Are you getting these *with* the concurrent fixpoint or without?


On Wed, Jul 29, 2015 at 2:18 PM, Chris Tetreault <notifications@github.com>
wrote:

> I just pulled liquidhaskell and liquid-fixpoint, ensured I had master
> checked out, and blew away and re-initialized my cabal sandboxes. I'm still
> getting these same failures.
>
> Perhaps I should open a separate issue for this in liquidhaskell?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquid-fixpoint/issues/88#issuecomment-126098240>
> .
>



-- 
Ranjit.

**Comment by ranjitjhala** [comment]
Also, what happens when you do:

    $ liquid tests/pos/ite1.hs

?

**Comment by ranjitjhala** [comment]
btw,

    $ liquid tests/pos/ite1.hs

passes fine with (master) on liquid-fixpoint AND (master) on LH.

(Btw, for this reason, it may be useful to have submodules...)


On Wed, Jul 29, 2015 at 2:28 PM, Ranjit Jhala <jhala@cs.ucsd.edu> wrote:

> Are you getting these *with* the concurrent fixpoint or without?
>
>
> On Wed, Jul 29, 2015 at 2:18 PM, Chris Tetreault <notifications@github.com
> > wrote:
>
>> I just pulled liquidhaskell and liquid-fixpoint, ensured I had master
>> checked out, and blew away and re-initialized my cabal sandboxes. I'm still
>> getting these same failures.
>>
>> Perhaps I should open a separate issue for this in liquidhaskell?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/ucsd-progsys/liquid-fixpoint/issues/88#issuecomment-126098240>
>> .
>>
>
>
>
> --
> Ranjit.
>



-- 
Ranjit.

**Comment by christetreault** [comment]
This is both with and without the concurrent fixpoint.

Using the master branch fixpoint:

```
chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ git pull
Already up-to-date.
chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ git checkout master
Already on 'master'
Your branch is up-to-date with 'origin/master'.
chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ cabal clean
cleaning...
chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ cabal install
Resolving dependencies...
In order, the following will be installed:
liquid-fixpoint-0.3.0.1 (reinstall)
Warning: Note that reinstalls are always dangerous. Continuing anyway...
Notice: installing into a sandbox located at
/home/chris/dmp/ucsd/liquid/ref-fixpoint/.cabal-sandbox
Configuring liquid-fixpoint-0.3.0.1...
Building liquid-fixpoint-0.3.0.1...
Installed liquid-fixpoint-0.3.0.1
chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ which fixpoint
/home/chris/dmp/ucsd/liquid/ref-fixpoint/.cabal-sandbox/bin/fixpoint
```

...and the master branch liquid:

```
chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ cd ../ref-liquid/
chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ git pull
Already up-to-date.
chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ git checkout master
M	.ghci
Already on 'master'
Your branch is up-to-date with 'origin/master'.
chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ cabal clean
cleaning...
chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ cabal install
[1 of 1] Compiling Main             ( dist/setup/setup.hs, dist/setup/Main.o )
Linking ./dist/setup/setup ...
Resolving dependencies...
In order, the following will be installed:
liquidhaskell-0.4.1.1 (reinstall)
Warning: Note that reinstalls are always dangerous. Continuing anyway...
Notice: installing into a sandbox located at
/home/chris/dmp/ucsd/liquid/ref-liquid/.cabal-sandbox
Configuring liquidhaskell-0.4.1.1...
Building liquidhaskell-0.4.1.1...
Installed liquidhaskell-0.4.1.1
chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ which liquid
/home/chris/dmp/ucsd/liquid/ref-liquid/.cabal-sandbox/bin/liquid
```

The result of `liquid tests/pos/ite1.hs`:

```
chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ liquid tests/pos/ite1.hs 
LiquidHaskell Copyright 2009-15 Regents of the University of California. All Rights Reserved.


**** DONE:  Extracted Core using GHC *******************************************
 

**** DONE:  generateConstraints ************************************************
 

**** START: fixpoint ***********************************************************
 


========================================================
© Copyright 2009 Regents of the University of California.
All Rights Reserved.
========================================================
syntax error at tests/pos/.liquid/ite1.hs.fq: 3.51
Fatal error: exception Parsing.Parse_error

**** DONE:  fixpoint ***********************************************************
 

**** DONE:  solve **************************************************************
 

**** DONE:  annotate ***********************************************************
 

**** ERROR: Unknown Error ******************************************************
```

I've uploaded the `ite1.hs.fq` to my google drive: https://drive.google.com/open?id=0Bx6BpheCXcBWOWpIMVJCanEwZ28

Unfortunately, drive thinks it's a binary file so you'll have to download it and open it in an editor.

**Comment by ranjitjhala** [comment]
This is almost surely because you have a stale "fixpoint.native" file,
which unfortunately is made worse by the whole cabal-sandbox business.

Can you (either use stack) or,

1. do 'which fixpoint.native' to find and delete it from wherever it is now,

2. go back and rebuild?

But frankly, its probably a lot easier to just use "stack"  -- I have a set
of instructions here:

https://github.com/ucsd-progsys/liquidhaskell#how-to-clone-build-and-install


On Wed, Jul 29, 2015 at 3:52 PM, Chris Tetreault <notifications@github.com>
wrote:

> This is both with and without the concurrent fixpoint.
>
> Using the master branch fixpoint:
>
> chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ git pull
> Already up-to-date.
> chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ git checkout master
> Already on 'master'
> Your branch is up-to-date with 'origin/master'.
> chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ cabal clean
> cleaning...
> chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ cabal install
> Resolving dependencies...
> In order, the following will be installed:
> liquid-fixpoint-0.3.0.1 (reinstall)
> Warning: Note that reinstalls are always dangerous. Continuing anyway...
> Notice: installing into a sandbox located at
> /home/chris/dmp/ucsd/liquid/ref-fixpoint/.cabal-sandbox
> Configuring liquid-fixpoint-0.3.0.1...
> Building liquid-fixpoint-0.3.0.1...
> Installed liquid-fixpoint-0.3.0.1
> chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ which fixpoint
> /home/chris/dmp/ucsd/liquid/ref-fixpoint/.cabal-sandbox/bin/fixpoint
>
> ...and the master branch liquid:
>
> chris@kittymitty:~/dmp/ucsd/liquid/ref-fixpoint$ cd ../ref-liquid/
> chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ git pull
> Already up-to-date.
> chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ git checkout master
> M   .ghci
> Already on 'master'
> Your branch is up-to-date with 'origin/master'.
> chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ cabal clean
> cleaning...
> chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ cabal install
> [1 of 1] Compiling Main             ( dist/setup/setup.hs, dist/setup/Main.o )
> Linking ./dist/setup/setup ...
> Resolving dependencies...
> In order, the following will be installed:
> liquidhaskell-0.4.1.1 (reinstall)
> Warning: Note that reinstalls are always dangerous. Continuing anyway...
> Notice: installing into a sandbox located at
> /home/chris/dmp/ucsd/liquid/ref-liquid/.cabal-sandbox
> Configuring liquidhaskell-0.4.1.1...
> Building liquidhaskell-0.4.1.1...
> Installed liquidhaskell-0.4.1.1
> chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ which liquid
> /home/chris/dmp/ucsd/liquid/ref-liquid/.cabal-sandbox/bin/liquid
>
> The result of liquid tests/pos/ite1.hs:
>
> chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ liquid tests/pos/ite1.hs
> LiquidHaskell Copyright 2009-15 Regents of the University of California. All Rights Reserved.
>
>
> **** DONE:  Extracted Core using GHC *******************************************
>
>
> **** DONE:  generateConstraints ************************************************
>
>
> **** START: fixpoint ***********************************************************
>
>
>
> ========================================================
> © Copyright 2009 Regents of the University of California.
> All Rights Reserved.
> ========================================================
> syntax error at tests/pos/.liquid/ite1.hs.fq: 3.51
> Fatal error: exception Parsing.Parse_error
>
> **** DONE:  fixpoint ***********************************************************
>
>
> **** DONE:  solve **************************************************************
>
>
> **** DONE:  annotate ***********************************************************
>
>
> **** ERROR: Unknown Error ******************************************************
>
> I've uploaded the ite1.hs.fq to my google drive:
> https://drive.google.com/open?id=0Bx6BpheCXcBWOWpIMVJCanEwZ28
>
> Unfortunately, drive thinks it's a binary file so you'll have to download
> it and open it in an editor.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquid-fixpoint/issues/88#issuecomment-126120044>
> .
>



-- 
Ranjit.

**Comment by christetreault** [comment]
I deleted `fixpoint.native` from `ref-fixpoint/.cabal-sandbox/bin` and `ref-liquid/.cabal-sandbox/bin` and then cleaned and rebuild both fixpoint and liquid. (which returns the one in `ref-fixpoint/.cabal-sandbox/bin`)

`fixpoint.native` rebuilt itself in `ref-fixpoint/.cabal-sandbox/bin` but not `ref-liquid/.cabal-sandbox/bin`. Still no change in behavior. I am able to run liquid on other files though.

I suppose I could try stack. It might take me a bit to get everything set up correctly. Though I imagine that this should work using cabal, since that is what any given user can be expected to have.

Prehaps GHC 7.10 is the culprit?

**Comment by christetreault** [comment]
Just for kicks and giggles, I copied fixpoint and fixpoint.native into the same folder as the liquid exectubale, and deleted them from their previous location so that:

```
chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ which liquid
/home/chris/dmp/ucsd/liquid/ref-liquid/.cabal-sandbox/bin/liquid
chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ which fixpoint
/home/chris/dmp/ucsd/liquid/ref-liquid/.cabal-sandbox/bin/fixpoint
chris@kittymitty:~/dmp/ucsd/liquid/ref-liquid$ which fixpoint.native
/home/chris/dmp/ucsd/liquid/ref-liquid/.cabal-sandbox/bin/fixpoint.native
```

No change in behavior.

**Comment by ranjitjhala** [comment]
No I am running ghc 710 and Travis tests with both

**Comment by ranjitjhala** [comment]
The culprit is probably that fixpoint.native is NOT getting rebuilt after you clean.

Can you wipe the slate clean (pushe your commits to github) then just start afresh with NEW repos?

**Comment by ranjitjhala** [comment]
Anyway if you can make progress with the OTHER LH benchmarks then do that and we can fix this when we meet next.

**Comment by christetreault** [comment]
OK, I'll do that tomorrow and let you know how it goes

**Comment by christetreault** [comment]
Today I blew away my local copies of the fixpoint and liquidhaskell repos so that `sudo find \ -name fixpoint*` returns no results.

I re-cloned and rebuild both, and tested `liquid tests/pos/ite1.hs`, however I am still seeing this same behavior.

I'm going to try it on my laptop next to see if it's some problem with my computer. I'll let you know how it goes.

**Comment by christetreault** [comment]
I'm seeing the same issue on my laptop, so either I'm doing something wrong, or my specific configuration has issues.

I'm going to back burner this for now; there's plenty of other stuff I can be doing in the meantime.

**Comment by gridaphobe** [comment]
@christetreault are you building liquid-fixpoint with `-fbuild-external`? Without that flag it will use the prebuilt binaries, which are usually only updated just-in-time for a release to hackage.

**Comment by christetreault** [comment]
@gridaphobe I was not, and that was the issue. Thanks!

**Comment by christetreault** [comment]
OK, so tests have been rerun on the `cores` branch. All tests pass using the single threaded implementation. All tests except the 5 listed in the issue heading pass using the parallel implementation.

**Comment by christetreault** [comment]
should `partition'` ever return the empty list?

**Comment by ranjitjhala** [comment]
No (unless he input is empty ie has no constraints)

**Comment by christetreault** [comment]
In that case, shouldn't it just return a singleton list with the same empty FInfo?

**Comment by ranjitjhala** [comment]
Yes you are right. 

Which test does this happen on?

**Comment by christetreault** [comment]
Let me rerun the tests to get you a list.

**Comment by christetreault** [comment]
I've opened a new issue (#93) for this empty list situation. It seems that this issue is much more pervasive than I initially thought.

**Comment by ranjitjhala** [comment]
Is this issue still open? or fixed by the above?

**Comment by christetreault** [comment]
This issue is still open

**Comment by christetreault** [comment]
I've spent the last few hours investigating this issue. It seems that for all these files, the function `wfKvars :: F.WfC a -> [F.KVar]` is being passed a `WfC a` that doesn't contain a `KVar`. Rather they all have some other `Pred` constructor as the result of `F.sr_reft`:


   * pos/Holes.hs:
```
PAtom Gt (EVar "y") (ECon (I 0))
```
   * pos/absref-crash0.hs:
```
PBexp (EApp "papp2" defined from: "../liquidhaskell/tests/pos/.liquid/absref-crash0.hs.fq" (line 1825, column 26) to: "../liquidhaskell/tests/pos/.liquid/absref-crash0.hs.fq" (line 1825, column 31) [EVar "p",EVar "VV",EVar "lq_tmp_x_235"])
```
   * pos/alphaconvert-List.hs:
```
PAtom Eq (EVar "v") (EApp "maxs" defined from: "../liquidhaskell/tests/pos/.liquid/alphaconvert-List.hs.fq" (line 15952, column 25) to: "../liquidhaskell/tests/pos/.liquid/alphaconvert-List.hs.fq" (line 15952, column 29) [EVar "lq_tmp_x_711"])
```
   * pos/LocalTermExpr.hs:
```
PAnd [PAtom Ge (EVar "lq_tmp_x_251") (ECon (I 0)),PBexp (EApp "Prop" defined from: "../liquidhaskell/tests/pos/.liquid/LocalTermExpr.hs.fq" (line 3520, column 55) to: "../liquidhaskell/tests/pos/.liquid/LocalTermExpr.hs.fq" (line 3520, column 59) [EVar "b"])]

PBexp (EApp "Prop" defined from: "../liquidhaskell/tests/pos/.liquid/LocalTermExpr.hs.fq" (line 3520, column 55) to: "../liquidhaskell/tests/pos/.liquid/LocalTermExpr.hs.fq" (line 3520, column 59) [EVar "b"])
```
   * pos/LocalHole.hs:
```
PAtom Le (EVar "v") (EVar "n#a18B")

PAtom Ge (EVar "v") (ECon (I 0))

PAnd [PAtom Ge (EVar "v") (ECon (I 0)),PAtom Le (EVar "v") (EVar "n#a18B")]
```

Links to my versions of the .fq files:

[Holes.hs.fq](https://gist.github.com/christetreault/8463f40efe74a2bde7ea)
[absref-crash.hs.fq](https://gist.github.com/christetreault/b4d09999c210a9dd27d7)
[alphaconvert-List.hs.fq](https://gist.github.com/christetreault/a67d2cce645f9047e8fc)
[LocalTermExpr.hs.fq](https://gist.github.com/christetreault/da5999b66a4d3e35b530)
[LocalHole.hs.fq](https://gist.github.com/christetreault/a6c743988318e3c089c3)

Would it be correct to assume that at this point all `WfC a` values that might result in this situation should be filtered out of the input?

**Comment by ranjitjhala** [comment]
Yes! (Your assumption is correct -- am surprised this was not done before...) thanks!

**Comment by christetreault** [comment]
OK, so based on @ranjitjhala's confirmation of my suspicions, I've pushed a change to cores that I believe fixes this issue:

[`wfGroup`](https://github.com/ucsd-progsys/liquid-fixpoint/blob/cores/src/Language/Fixpoint/Partition.hs#L176-L178) Now returns a `Maybe Int`, rather than throwing an exception.

[`maybeInserts`](https://github.com/ucsd-progsys/liquid-fixpoint/blob/cores/src/Language/Fixpoint/Misc.hs#L160-L163) In `Misc.hs` is a version of inserts that takes a `Maybe Int`, and if `Nothing`, just returns the original `Map`

[`maybeGroupMap`](https://github.com/ucsd-progsys/liquid-fixpoint/blob/cores/src/Language/Fixpoint/Misc.hs#L183) Is a version of `groupMap` that calls `maybeInserts` instead of `inserts`.

[`partitionByConstraints`](https://github.com/ucsd-progsys/liquid-fixpoint/blob/cores/src/Language/Fixpoint/Partition.hs#L156) updated to use `maybeGroupMap`.

The end result is that if `wfGroup` gets a `WfC` that contains something that isn't a `KVar`, it gets dropped. All tests are now passing, and if management approves of this solution I believe that this issue can be closed.

**Comment by ranjitjhala** [comment]
Looks good! Thanks @ctetreau.

On Fri, Aug 21, 2015 at 10:26 AM, Chris Tetreault <notifications@github.com>
wrote:

> OK, so based on @ranjitjhala <https://github.com/ranjitjhala>'s
> confirmation of my suspicions, I've pushed a change to cores that I believe
> fixes this issue:
>
> wfGroup
> <https://github.com/ucsd-progsys/liquid-fixpoint/blob/cores/src/Language/Fixpoint/Partition.hs#L176-L178>
> Now returns a Maybe Int, rather than throwing an exception.
>
> maybeInserts
> <https://github.com/ucsd-progsys/liquid-fixpoint/blob/cores/src/Language/Fixpoint/Misc.hs#L160-L163>
> In Misc.hs is a version of inserts that takes a Maybe Int, and if Nothing,
> just returns the original Map
>
> maybeGroupMap
> <https://github.com/ucsd-progsys/liquid-fixpoint/blob/cores/src/Language/Fixpoint/Misc.hs#L183>
> Is a version of groupMap that calls maybeInserts instead of inserts.
>
> partitionByConstraints
> <https://github.com/ucsd-progsys/liquid-fixpoint/blob/cores/src/Language/Fixpoint/Partition.hs#L156>
> updated to use maybeGroupMap.
>
> The end result is that if wfGroup gets a WfC that contains something that
> isn't a KVar, it gets dropped. All tests are now passing, and if
> management approves of this solution I believe that this issue can be
> closed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquid-fixpoint/issues/88#issuecomment-133504256>
> .
>



-- 
Ranjit.
